File: 021_Merge_Two_Sorted_Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(0)
        thisNode = result
        while l1 and l2:
            if l1.val < l2.val:
                thisNode.next = l1
                l1 = l1.next
            else:
                thisNode.next = l2
                l2 = l2.next
            thisNode = thisNode.next
        thisNode.next = l1 if l1 else l2
        return result.next
    

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = tail = ListNode()
        while list1 and list2 :
            if list1.val < list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next

        if list1:
            tail.next = list1
        if list2:
            tail.next = list2
        
        return dummy.next

File: 021_Merge_Two_Sorted_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(0);
        ListNode* cur = dummy;
        dummy->next = l1;
        while (l1 && l2) {
            if (l1->val < l2->val)
                l1 = l1->next;
            else {
                ListNode* tmp = l2->next;
                cur->next = l2;
                l2->next = l1;
                l2 = tmp;
            }
            cur = cur->next;
        }
        if (l2) cur->next = l2;
        return dummy->next;
    }
};
